### PR TITLE
Revert conduit 3 changes

### PR DESCRIPTION
**Pull request by samoht**
As we are delaying the release of conduit 3.0 again (because we are still not fully happy with the complexity added by the new API) I am temporary reverting https://github.com/mirage/ocaml-cohttp/pull/692. As a few changes have been made to master since then, I had to do it manually. Hopefully I haven't broken anything.

/cc @dinosaure and @lyrm

### CODE REVIEW

**Comment by avsm** [comment]
I'm fine with this. We'll revisit the conduit changes (which were very much heading in the right direction) once we have the maintainer bandwidth to figure out the revdeps cost to the community.  Since cohttp 3.x has been tagged already, we cannot re-release, so lets bump to cohttp 4 for future API changes, and conduit2 can come back in cohttp 5.

@samoht we need to make the CI pass. I can take a quick look at that in a few hours and push to your branch.

@mseri -- this ok with you?

**Comment by samoht** [comment]
I'll fix the CI.

I also haven't realised that cohttp 3 was released (but it has `available: false` in its opam metadata so I guess nobody is using it yet?)

**Comment by avsm** [comment]
That's right -- it got marked as available:false fairly soon after because of the revdeps changes, but it's there. Unfortunately the new package names will show up as a result on searches. We'll need to edit the opam metadata.

**Comment by mseri** [comment]
I think it will be hard to cherrypick the fixes that went in during the preparation for 3.0.0 and would be a pity to lose them. Why don't we just wait and update the code once the new conduit is released? 

Do we really need to sync the version numbers of conduit and cohttp? 3.0.0 is tagged but not released... we could always tag 3.0.1 and release that as the official 3.0.0 once appropriate. 

(In fact, since it is unavailable on opam-repository, we could really remove it there -- still preserving the 3.0.0 tag and release here unofficially). Or release 3.0.0~alpha and leave it there for people that want to try the current API, with a huge warning that there are further upcoming breaking changes.

**Comment by samoht** [comment]
> I think it will be hard to cherrypick the fixes that went in during the preparation for 3.0.0 and would be a pity to lose them. Why don't we just wait and update the code once the new conduit is released?

Nobody has time right now to focus on the new conduit, so let's not block ourselves on this and let's continue to improve (and release) cohttp. For instance @lyrm is interested to contribute to improve cohttp headers, so my motivation is to unblock her asap :-)

Also while doing this I noticed a few clean-up that I'd like to do next (use ocamlformat, move src and tests around).

**Comment by samoht** [comment]
(also I'll double check but I think I've cherry-picked everything that needed to be cherry-picked :p)

**Comment by mseri** [comment]
In that case, go ahead!

**Comment by mseri** [comment]
I think we can release this as 3.0.1 no matter what, keeping 3.0.0 unavailable or completely removing it from opam.

**Comment by mseri** [comment]
I think we will need to fix the server examples in the README, we should probably use mdx to compile them and keep them updated (I was doing it manually by also having them in the examples folder)

**Comment by samoht** [comment]
> I think it will be hard to cherrypick the fixes that went in during the preparation for 3.0.0 and would be a pity to lose them.

Ok I actually missed a few renaming commits. Fixing that now.

**Comment by avsm** [comment]
I'm fine with this being 3.x, but make it 3.1.0 at least. It really isn't a point release over what 3.0.0 was tagged as ;-)

**Comment by mseri** [comment]
I have no preference really, just wanted to avoid skipping 3 since it is effectively unreleased :P
It's the first uneven prime!

**Comment by avsm** [comment]
Skipping the atomic number of lithium does seem like a pity! Let's go with 3.1.0 @samoht

**Review by mseri — APPROVED** [review]
Can you also fix the two linting issues in the CI?

**Comment by samoht** [comment]
Ok I think I've cherry-picked everything that I could do now. Please check that the diff looks ok.

The one that makes me very sad is reverting support for client-side certificates. I am so sad that I might make a small patch to conduit to restore this...

**Comment by samoht** [comment]
Ok all fixed now -- I think it's ready to be merged once everyone is happy about the revert.

File: cohttp-lwt-unix/src/server.mli
Comment (mseri) [inline cohttp-lwt-unix/src/server.mli:18]:
```suggestion
    and how [conduit-unix] is configured.
```

**Review by mseri — APPROVED** [review]
LGTM

**Comment by samoht** [comment]
The [client-side CA cert patch](https://github.com/mirage/ocaml-cohttp/pull/692/commits/8443defbf9dafc9cc2b82352e9806c2e143345d5 ) has been backported to conduit 2: https://github.com/mirage/ocaml-conduit/pull/374

**Comment by mseri** [comment]
Fantastic! I went over the changes again, it looks fine to me. @avsm do you also give the green light for a merge?

File: CHANGES.md
Comment (avsm) [inline CHANGES.md:8]:
```suggestion
- Revert the changes to adapt to conduit 3.0.0 that were briefly present in cohttp.3.0.0.  We will revisit these in a future release of cohttp, but for now have chosen to preserve better compatibility with existing cohttp users. (#741, @samoht)
```

**Review by avsm — APPROVED** [review]
Just a minor expansion to the CHANGES file to explain the PR

**Pull request merged**
